src/base/bittorrent/session.cpp

```diff
@@ -1887,8 +1889,58 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
 
         p = magnetUri.addTorrentParams();
     }
-    else if (torrentInfo.isValid()) {
-        if (!params.restored) {
+    else {
+        if (!torrentInfo.isValid()) {
+            // We can have an invalid torrentInfo when there isn't a matching
+            // .torrent file to the .fastresume we loaded. Possibly from a
+            // failed upgrade.
+            return false;
+        }
+
+        hash = torrentInfo.hash();
+    }
+
+    // We should not add the torrent if it is already
+    // processed or is pending to add to session
+    if (m_addingTorrents.contains(hash) || m_loadedMetadata.contains(hash))
+        return false;
+
+    TorrentHandle *const torrent = m_torrents.value(hash);
+    if (torrent) {  // a duplicate torrent is added
+        if (torrent->isPrivate() || (!fromMagnetUri && torrentInfo.isPrivate()))
+            return false;
+
+        // merge trackers and web seeds
+        torrent->addTrackers(fromMagnetUri ? magnetUri.trackers() : torrentInfo.trackers());
+        torrent->addUrlSeeds(fromMagnetUri ? magnetUri.urlSeeds() : torrentInfo.urlSeeds());
+        return true;
+    }
+
+    // Detect if fastresume contains downloaded metadata
+    // In the case of magent metadata isn't fetched yet, we still need
+    // to setup the default parameters.
+    bool isMetadataAvailable = false;
+
+    if (!fromMagnetUri) {
+        if (params.restored) {  // load from existing fastresume
+#if (LIBTORRENT_VERSION_NUM < 10200)
+            p.resume_data = std::vector<char> {fastresumeData.constData()
+                , (fastresumeData.constData() + fastresumeData.size())};
+            p.flags |= lt::add_torrent_params::flag_use_resume_save_path;
+
+            // we still setup the default parameters and
+            // let libtorrent handle the parameter merging
+            isMetadataAvailable = false;
+#else
+            lt::error_code ec;
+            p = lt::read_resume_data(fastresumeData, ec);
+
+            // fastresume will only contain "qBt-*" entries when metadata
+            // hasn't been downloaded
+            isMetadataAvailable = !p.file_priorities.empty();
```

Comment:
Here is the case when metadata is obviously available (since we have .torrent file). We are under `!fromMagnetUri` condition here. But we have no valid "fastresume" data (i.e. stored torrent progress and settings). So I'm saying "isMetadataAvailable" is not the right variable name. Unavailable metadata is just one of the reasons why we have no full fastresume data when we load torrent.

Reply:
>Here is the case when metadata is obviously available (since we have .torrent file). 

I still don't think we are seeing the same thing, I see a .fastresume file but no .torrent file.
You can observe it by adding a fake hash to download (that metadata retrieval will never succeed) and then restart qbt.

>Unavailable metadata is just one of the reasons why we have no full fastresume data when we load torrent.

I see we have different definition of *valid* fastresume data. My definition is just simple, if there exist such file and adheres to bencoding then it is valid. But your seems more than that...

Reply:
>I see a .fastresume file but no .torrent file.

`params.restored == true` means that we have .fastresume file
`fromMagnetUri == false` means we have .torrent file (i.e. we have metadata)
Unless I missed/forgo something...
>But your seems more than that...

Yes. Here (in this part of review) I refer to the libtorrent-generated part of data containing in .fastresume file. I.e. if this data is unavailable (new torrent, or restored torrent having no such data for some reason) we should load default values.

Reply:
OFFTOPIC.
Actually I define fastresume data as the data that allows to restore torrent without full recheck. All other is torrent settings.
The problem is that the libtorrent itself mixes them. Therefore, it does not generate fastresume data when metadata is not available, so real "fastresume" data (torrent progress etc.) is also not available (because it does not make sense), but all other torrent settings still make sense.
I think .fastresume file is incorrect name. But I don't propose to change it because of some compatibility issues between qBittorrent versions.

Reply:
>>I see a .fastresume file but no .torrent file.
>
>params.restored == true means that we have .fastresume file
fromMagnetUri == false means we have .torrent file (i.e. we have metadata)
Unless I missed/forgo something...

`fromMagnetUri` comes from `magnetUri.isValid()` so I don't think it is 100% related to having .torrent (metadata) file.
I didn't trace the code outside of this function, so I have little idea what is going on out there. 

If you have time to start qbt, please do try the fake hash mentioned in previous post, you will find that "no .torrent file" condition (no metadata) is true.

Reply:
>you will find that "no .torrent file" condition (no metadata) is true.

I know it. That doesn't contradict my statements above.

Reply:
I really like to avoid complicating the definition of valid fastresume (thus rejecting `isFastresumeAvailable`), so would you accept alternatives such as: `isFastresumeComplete` or `isFastresumeUsable`?

Reply:
For the most part, I objected to using "metadata" here.
"isFastresumeComplete" could be ok for me (or "hasCompleteFastresume").

Reply:
renamed to `hasCompleteFastresume`.